### PR TITLE
Add configurable query timeout

### PR DESCRIPTION
**Pull request by edevil**
Addresses #1713 
Based on work from @juliusza

### CODE REVIEW

File: test/integration/client/api-tests.js [hunk 1]
```diff
@@ -15,6 +15,17 @@ suite.test('pool callback behavior', done => {
   })
 })
 
+suite.test('query timeout', (cb) => {
+  const pool = new pg.Pool({query_timeout: 1000})
+  pool.connect().then((client) => {
+    client.query('SELECT pg_sleep(2)', assert.calls(function (err, result) {
+      assert(err)
```
Comment (kibertoad) [inline test/integration/client/api-tests.js]: can we assert what kind of error we get?

File: test/integration/client/api-tests.js [hunk 2]
```diff
@@ -15,6 +15,28 @@ suite.test('pool callback behavior', done => {
...
+      client.release()
+      pool.end(cb)
+    }))
+  })
+})
+
+suite.test('query timeout', (cb) => {
```
Comment (kibertoad) [inline test/integration/client/api-tests.js]: is name correct?

Comment (edevil) [reply test/integration/client/api-tests.js]: Right, I'll change it right away. Thanks.

File: test/integration/client/api-tests.js (hunk 1 above)
Comment (edevil) [reply test/integration/client/api-tests.js]: i'll check the message.

Comment (edevil) [reply test/integration/client/api-tests.js]: @kibertoad done

File: test/integration/client/api-tests.js (hunk 2 above)
Comment (edevil) [reply test/integration/client/api-tests.js]: @kibertoad done

File: lib/client.js [hunk 3]
```diff
@@ -416,6 +421,40 @@ Client.prototype.query = function (config, values, callback) {
...
+      if (index > -1) {
+        this.queryQueue.splice(index, 1)
+      }
+
+      this._pulseQueryQueue()
+    }, readTimeout)
+
+    query.callback = (err, res) => {
+      clearTimeout(readTimeoutTimer)
+      queryCallback(err, res)
```
Comment (joelmukuthu) [inline lib/client.js]: Not sure, but should there be a check for `typeof queryCallback` before making this call (seeing as there was one on line 434)?

Comment (edevil) [reply lib/client.js]: Yes, I don't think it's needed after all. I've removed it.

**Comment by juliusza** [comment]
Nice! I have briefly tested my code with query timeout enabled in production and it seemed to work fine.

Eventually I found that the issue was caused by misconfigured tcp proxy, and decided not to pursue this. Regardless, this could be very useful to recover stuck TCP sessions due to lossy network conditions @brianc. Also this is completely optional, and the lib will function as it did without the setting enabled.

**Comment by edevil** [comment]
@juliusza Is that documentation in some other repository?

**Comment by abenhamdine** [comment]
> @juliusza Is that documentation in some other repository?  

Unfortunately, so far the docs are not available for contributions.

**Comment by brianc** [comment]
> Unfortunately, so far the docs are not available for contributions.

I'm thinking of porting them to [https://www.gitbook.com] so it's easier to contribute.

**Comment by edevil** [comment]
@brianc what about this PR?

**Comment by brianc** [comment]
Have the flu in bed. Will check as soon as im back in my feet. Havenf
forgotten

**Comment by abenhamdine** [comment]
> Have the flu in bed. 

Winter is coming :mask:

**Comment by edevil** [comment]
@brianc how’s that flu coming along? :)

**Comment by brianc** [comment]
The flu suuuuuuuuuuuuuuucked!  Thanks for asking. :smile: It cleared up early last week & now I just have left-over bronchitis.  Definitely going to get the flu shot from here on out.  Now...lemme look at this PR! Sorry for the review (thanks for pinging me gently on it!)

**Comment by abenhamdine** [comment]
> This is great, thanks! I'll get this merged & push out a new minor version today. heart

That's great :+1: Could you also bump the version of pg-pool to 2.0.4, in order to be able to benefit from https://github.com/brianc/node-pg-pool/commit/1871d0f9e18330afe8149ff5df8b3e54ac71469a ? thx !

**Comment by brianc** [comment]
@abenhamdine you bet!

**Comment by neylsongularte** [comment]
I'm looking forward to this functionality

**Comment by edevil** [comment]
@brianc Hello! Any news on this?

**Comment by kibertoad** [comment]
@brianc Is pg-pool bump also happening?

**Comment by brianc** [comment]
@kibertoad yah definitely, right after this

published `pg@7.7.0`

**Comment by brianc** [comment]
pg-pool is already published at `2.0.4` as of a [few weeks ago](https://github.com/brianc/node-pg-pool/commit/4b9669eaa764126135f0bbf64af9be1e9d5092a9).  Am I missing something / are you not able to install it?

**Comment by kibertoad** [comment]
@brianc It's listed in package.json as `"pg-pool": "~2.0.3"` which means that projects with package-lock.json are not going to pick it up unless you regenerate entire lock file.

**Comment by brianc** [comment]
ohhh duh yeah i can fix that no problem, sorry

**Comment by brianc** [comment]
should I just change it to `pg-pool:^2.0.3`?  I think that'd be better.

**Comment by brianc** [comment]
hmmm wait the `~2.0.3` doesn't prevent you from installing `2.0.4` it just doesn't upgrade it if you already have 2.0.3 installed.  K, i'll just bump it.

**Comment by brianc** [comment]
K published `pg@7.7.1` bumping the min version for pg-pool to `2.0.4`

**Comment by abenhamdine** [comment]
great, thx @brianc !

**Comment by kibertoad** [comment]
Thank you!

**Comment by jfromaniello** [comment]
Does this actually cancel the query in the server once it timedout in the client side? isn't clear from the diff.

Thank you!

**Comment by juliusza** [comment]
Hey @jfromaniello, this does not cancel query on the server. It would be reasonable to configure both query_timeout and statement_timeout, making the statement_timeout slightly longer so that it times out right after the client stopped waiting for response.

**Comment by jfromaniello** [comment]
@juliusza thank you very much!

**Comment by johan13** [comment]
How is the documentation for this feature coming?

The `query_timeout` property is missing from the typescript typings, and the typings seem to be based on the docs, so I thought I'd start here before I put in a merge request to [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/pg).

**Comment by brianc** [comment]
I recently open-sourced the docs!
https://github.com/brianc/node-postgres-docs . so feel free to add any
documentation you want over there!

**Comment by tmtron** [comment]
> Hey @jfromaniello, this does not cancel query on the server. It would be reasonable to configure both query_timeout and statement_timeout, making the statement_timeout slightly longer so that it times out right after the client stopped waiting for response.

Hm.. I thought it should be the other way around: the query-timeout should be slightly longer than the statement-timeout.
i.e. when the statement-timeout is reached, the database will cancel the query and return the [error ](https://www.postgresql.org/docs/11/errcodes-appendix.html) `57014 query_canceled` to the client: so the client gets the error-details and knows exactly what happened.
If, the server cannot return any result/error back to the client for some reason, then the query-timeout will kick in - and just stop listening for the server result.

@juliusza what do you think?



`

**Comment by juliusza** [comment]
Yes @tmtron, make sense to have `query-timeout` longer than `statement-timeout`.

`[CRITICAL] TypeError: Cannot read property 'handleCommandComplete' of null
at Connection.<anonymous> (/opt/acn/release_20190925/api/node_modules/pg/lib/client.js:325:22)`

I found that sometimes I get this error on production. Because pg lib is event based, there's no proper stack trace to work with. Obviously we haven't thought of a certain edge case here, that causes above error to crash node process.

I no longer need the query-timeout setting, so I don't think I'll want to spend time working on a fix.

**Pull request merged**